hal/src/nRF52840/exflash_hal.c

```diff
@@ -388,64 +563,96 @@ int hal_exflash_copy_sector(uintptr_t src_addr, uintptr_t dest_addr, size_t data
-        return -1;
-    }
-
-    /* Validate bounds */
-    if (addr + data_size > MX25_OTP_SECTOR_SIZE) {
-        return -2;
-    }
+    CHECK_TRUE(sp == HAL_EXFLASH_SPECIAL_SECTOR_OTP, SYSTEM_ERROR_INVALID_ARGUMENT);
+    CHECK_TRUE(addr + data_size <= hal_exflash_driver.otp_size, SYSTEM_ERROR_OUT_OF_RANGE);
+    CHECK_TRUE(data_buf != NULL, SYSTEM_ERROR_INVALID_ARGUMENT);
+    int ret = 0;
 
     hal_exflash_lock();
 
-    /* Enter Secure OTP mode */
-    int ret = enter_secure_otp();
-    if (ret) {
-        ret = -3;
-        goto hal_exflash_read_special_done;
-    }
+    if (hal_exflash_driver.type == QSPI_TYPE_GD25WQ64EQFG) {
+        // Determine which OTP Register to start writing to
+        size_t otp_page = addr / GD25_OTP_SECTOR_SIZE;
+        size_t data_to_read = data_size;
+        size_t data_read = 0;
+        size_t read_amount = 0;
+
+        while (data_to_read && !ret) {
+            // Determine the offset of the current OTP register to start writing at
+            size_t register_offset = (addr + data_read) % GD25_SECURITY_REGISTER_SIZE;
+            read_amount = data_to_read;
+
+            // If the data remaining to read extends into the next register, truncate the read amount to fit into this register
+            if (register_offset + read_amount > GD25_SECURITY_REGISTER_SIZE){
+                read_amount = GD25_SECURITY_REGISTER_SIZE - register_offset;
+            }
+
+            ret = gd25_secure_register_read(otp_page, register_offset, &data_buf[data_read], read_amount);
 
-    ret = hal_exflash_read(addr, data_buf, data_size);
+            // Increment/decrement counter variables, increment to the next OTP register
+            data_to_read -= read_amount;
+            data_read += read_amount;
+            otp_page++;
+        }
+    } else {
+        ret = enter_secure_otp();
+        if (!ret) {
+            ret = hal_exflash_read(addr, data_buf, data_size);
+        }
+        exit_secure_otp(); 
+    }
 
-hal_exflash_read_special_done:
-    /* Exit Secure OTP mode */
-    exit_secure_otp();
     hal_exflash_unlock();
     return ret;
 }
 
 int hal_exflash_write_special(hal_exflash_special_sector_t sp, uintptr_t addr, const uint8_t* data_buf, size_t data_size) {
-    /* The only special sector we have on MX25L3233F is OTP */
-    if (sp != HAL_EXFLASH_SPECIAL_SECTOR_OTP) {
-        return -1;
-    }
-
-    /* Validate bounds */
-    if (addr + data_size > MX25_OTP_SECTOR_SIZE) {
-        return -1;
-    }
+    CHECK_TRUE(sp == HAL_EXFLASH_SPECIAL_SECTOR_OTP, SYSTEM_ERROR_INVALID_ARGUMENT);
+    CHECK_TRUE(addr + data_size <= hal_exflash_driver.otp_size, SYSTEM_ERROR_OUT_OF_RANGE);
+    CHECK_TRUE(data_buf != NULL, SYSTEM_ERROR_INVALID_ARGUMENT);
+    int ret = 0;
 
     hal_exflash_lock();
 
-    /* Enter Secure OTP mode */
-    int ret = enter_secure_otp();
-    if (ret) {
-        goto hal_exflash_write_special_done;
-    }
+    if (hal_exflash_driver.type == QSPI_TYPE_GD25WQ64EQFG) {
```

Comment:
The GD25 OTP register read/writes are copy/paste of each other basically. Could refactor/consolidate this if this isnt ok